[SPARK-35593][K8S][CORE] Support shuffle data recovery on the reused PVCs

dongjoon-hyun opened the pull request:
### What changes were proposed in this pull request?

Previously, the following two commits allow driver-owned on-demand PVC reuse.
- SPARK-35182 Support driver-owned on-demand PVC
- SPARK-35416 Support PersistentVolumeClaim Reuse

This PR aims to recover the shuffle data on those remounted PVCs. The lifecycle of PVCs are tied to the one of Spark jobs. Since this is K8s specific feature, `ShuffleDataIO` plugin is used.

### Why are the changes needed?

Although Pod is killed, we can remount PVCs and recover some data from it.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the newly added test cases.

dongjoon-hyun committed:
[SPARK-35593][K8S][CORE] Support shuffle data recovery on the reused PVCs

SparkQA commented:
**[Test build #139156 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139156/testReport)** for PR 32730 at commit [`44e7920`](https://github.com/apache/spark/commit/44e79202e1313873ad74b40fd1ae36d655e1a4b1).

SparkQA commented:
**[Test build #139156 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139156/testReport)** for PR 32730 at commit [`44e7920`](https://github.com/apache/spark/commit/44e79202e1313873ad74b40fd1ae36d655e1a4b1).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KubernetesLocalDiskShuffleDataIO(sparkConf: SparkConf) extends ShuffleDataIO `
  * `class KubernetesLocalDiskShuffleExecutorComponents(sparkConf: SparkConf)`

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139156/

SparkQA commented:
**[Test build #139159 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139159/testReport)** for PR 32730 at commit [`ce8d599`](https://github.com/apache/spark/commit/ce8d599b187e8c1564089dd8d670cf6a1d0a36b6).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43678/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43678/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43678/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43680/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43680/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43680/

SparkQA commented:
**[Test build #139159 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139159/testReport)** for PR 32730 at commit [`ce8d599`](https://github.com/apache/spark/commit/ce8d599b187e8c1564089dd8d670cf6a1d0a36b6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KubernetesLocalDiskShuffleDataIO(sparkConf: SparkConf) extends ShuffleDataIO `
  * `class KubernetesLocalDiskShuffleExecutorComponents(sparkConf: SparkConf)`

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139159/

mridulm commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -234,6 +252,7 @@ private class ShuffleStatus(
     for (mapIndex <- mapStatuses.indices) {
       if (mapStatuses(mapIndex) != null && f(mapStatuses(mapIndex).location)) {
         _numAvailableMapOutputs -= 1
+        mapStatusesDeleted(mapIndex) = mapStatuses(mapIndex)
```
This has potential for interacting badly with correctness changes, right ?
See `DAGScheduler.submitMissingTasks` when stage is Indeterminate

dongjoon-hyun replied:
In that case, `mapId` is not the same, isn't it? We are reusing with `mapId` at [line 170](https://github.com/apache/spark/pull/32730/files#diff-a3b15298f97577c1fadcc2d76d015eebd6343e246c6717417d33f3c458847f46R170), @mridulm .
```
val index = mapStatusesDeleted.indexWhere(x => x != null && x.mapId == mapId)
```

dongjoon-hyun replied:
BTW, I agree with you that we don't have a test coverage for the indeterministic stage case.

mridulm replied:
You are right, the check on mapId and mapIndex ensure correctness. Thanks for clarifying !

viirya commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -146,6 +151,7 @@ private class ShuffleStatus(
       _numAvailableMapOutputs += 1
       invalidateSerializedMapOutputStatusCache()
     }
+    mapStatusesDeleted(mapIndex) = mapStatuses(mapIndex)
```
This is `addMapOutput`, why we need to add it to `mapStatusesDeleted`?

mridulm replied:
Probably also remove from `mapStatusesDeleted` if present ?

dongjoon-hyun replied:
Hi, @viirya and @mridulm .
This is based on the semantic of this function description. With this, we will recover only the last one.
```
  /**
   * Register a map output. If there is already a registered location for the map output then it
   * will be replaced by the new location.
   */
```

mridulm replied:
(I am assuming @viirya has the same query)
Given we have a new map output, which we are storing as `mapStatuses(mapIndex) = status` in the next line, why are we saving the previous mapStatus ?
I am sure I am missing something here.

viirya replied:
yea, same question.

dongjoon-hyun replied:
Ah, I missed your intention. Right, AS-IS code is matching `mapStatusesDeleted` every places where `mapStatuses` is changed. I'll remove this part. Thanks, @viirya and @mridulm .

viirya commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -154,14 +160,25 @@ private class ShuffleStatus(
-          logWarning(s"Asked to update map output ${mapId} for untracked map status.")
+          val index = mapStatusesDeleted.indexWhere(x => x != null && x.mapId == mapId)
+          if (index >= 0) {
+            val mapStatus = mapStatusesDeleted(index)
+            mapStatus.updateLocation(bmAddress)
+            assert(mapStatuses(index) == null)
+            mapStatuses(index) = mapStatus
+            _numAvailableMapOutputs += 1
+            invalidateSerializedMapOutputStatusCache()
+            logDebug(s"Recover ${mapStatus.mapId} ${mapStatus.location}")
```
Do we need to clear `mapStatusesDeleted(index)`?

dongjoon-hyun replied:
No, we should not clear `mapStatusesDeleted` after this recovery. We may need to recover more than once if this executor dies again.

mridulm replied:
But that will trigger another `removeMapOutput` right ?
And so get saved in `mapStatusesDeleted` at that time ?

dongjoon-hyun replied:
Right, it will be matched at that time. I'll update the PR.

viirya commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleExecutorComponents.scala:
```diff
@@ -0,0 +1,105 @@
+    // This is not used.
+    val classTag = implicitly[ClassTag[Object]]
+    val level = StorageLevel.DISK_ONLY
+    val (indexFiles, dataFiles) = files.partition(_.getName.endsWith(".index"))
+    (dataFiles ++ indexFiles).foreach { f =>
```
I don't see `dataFiles` and `indexFiles` are treated differently below, do we need to partition them?

dongjoon-hyun replied:
Thank you for review, @viirya . Apache Spark assumes that `dataFile` arrives first.
- https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala#L599

viirya commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleExecutorComponents.scala:
```diff
@@ -0,0 +1,105 @@
+    if (sparkConf.getBoolean("spark.kubernetes.driver.reusePersistentVolumeClaim", false)) {
+      // Turn off the deletion of the shuffle data in order to reuse
+      blockManager.diskBlockManager.deleteFilesOnStop = false
+      logError("Recover shuffle data")
```
This doesn't always recover shuffle data, right? Once `KubernetesLocalDiskShuffleDataIO` is used, `initializeExecutor` will be called per executor, even it is initial executor without failure.

dongjoon-hyun replied:
Yep. Right. This is a left-over of my debugging message. I used `logError`. :)

dongjoon-hyun committed:
Remove logError

SparkQA commented:
**[Test build #139314 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139314/testReport)** for PR 32730 at commit [`b54de36`](https://github.com/apache/spark/commit/b54de3668e1443b14d4272cd47f664298b3cbb3e).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43836/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43836/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43836/

SparkQA commented:
**[Test build #139314 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139314/testReport)** for PR 32730 at commit [`b54de36`](https://github.com/apache/spark/commit/b54de3668e1443b14d4272cd47f664298b3cbb3e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139314/

dongjoon-hyun committed:
Address comments

SparkQA commented:
**[Test build #139343 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139343/testReport)** for PR 32730 at commit [`1efa54e`](https://github.com/apache/spark/commit/1efa54eb7393e84233430a81c181ff8d56554754).

dongjoon-hyun committed:
Clear mapStatusesDeleted after recover

SparkQA commented:
**[Test build #139344 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139344/testReport)** for PR 32730 at commit [`85f90f3`](https://github.com/apache/spark/commit/85f90f3fff13dafc2acfcffb7c37253e4087ecbc).

SparkQA commented:
**[Test build #139344 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139344/testReport)** for PR 32730 at commit [`85f90f3`](https://github.com/apache/spark/commit/85f90f3fff13dafc2acfcffb7c37253e4087ecbc).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139344/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43865/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43866/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43865/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43866/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43865/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43866/

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #139347 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139347/testReport)** for PR 32730 at commit [`85f90f3`](https://github.com/apache/spark/commit/85f90f3fff13dafc2acfcffb7c37253e4087ecbc).

SparkQA commented:
**[Test build #139343 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139343/testReport)** for PR 32730 at commit [`1efa54e`](https://github.com/apache/spark/commit/1efa54eb7393e84233430a81c181ff8d56554754).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43869/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43869/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139343/

SparkQA commented:
**[Test build #139347 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139347/testReport)** for PR 32730 at commit [`85f90f3`](https://github.com/apache/spark/commit/85f90f3fff13dafc2acfcffb7c37253e4087ecbc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139347/

mridulm commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -154,14 +159,26 @@ private class ShuffleStatus(
+          if (index >= 0) {
+            val mapStatus = mapStatusesDeleted(index)
+            mapStatus.updateLocation(bmAddress)
+            assert(mapStatuses(index) == null)
```
I am fine with the PR, just had one q:

I have not looked into decomissioning in detail - but this assumption will hold right ?
Can there be an interleaving recomputation or speculative task which updates MOT for that index (not mapId) ?

dongjoon-hyun replied:
> I have not looked into decomissioning in detail - but this assumption will hold right (assertion) ?

For the worker decommission, yes. There are three cases.
- When the worker is decommissioning, it should be handled at the previous `Some(mapStatus)`.
- When the worker is killed before finishing, new executor is started before handling the dead executors. It should be handled at the previous `Some(mapStatus)`.
- When the worker is killed before finishing, the data of the deleted executors are removed and `Asked to update map output ${mapId} for untracked map status`  happened before. And, this PR is fixing it.

For the following, if something is filled here already with the different mapId, we can update this or ignore this. However, in both case, `assert` is too strong. In that case, we had better skip in that case. Thanks! I'll update this PR.
> Can there be an interleaving recomputation or speculative task which updates MOT for that partition index ?

dongjoon-hyun committed:
Remove assert

dongjoon-hyun commented:
Thank you for review, @mridulm and @viirya . I addressed the comments and rebased this to the master to bring https://github.com/apache/spark/pull/32784 .

SparkQA commented:
**[Test build #139391 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139391/testReport)** for PR 32730 at commit [`a9bf099`](https://github.com/apache/spark/commit/a9bf099c653ccd485e3409dc20f9d067974deb53).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43913/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43913/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43913/

SparkQA commented:
**[Test build #139391 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139391/testReport)** for PR 32730 at commit [`a9bf099`](https://github.com/apache/spark/commit/a9bf099c653ccd485e3409dc20f9d067974deb53).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139391/

dongjoon-hyun commented:
Hi, @mridulm and @viirya . Could you review this once more?

viirya commented:
@dongjoon-hyun Yea, I'll do the review today. Thanks.

dongjoon-hyun commented:
Thank you so much, @viirya !

viirya commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleExecutorComponents.scala:
```diff
@@ -0,0 +1,102 @@
+          f.isDirectory && f.getName.startsWith("spark-")
+        }
+        val files = oldDirs
+          .flatMap(_.listFiles).filter(_.isDirectory) // executor-xxx
+          .flatMap(_.listFiles).filter(_.isDirectory) // blockmgr-xxx
+          .flatMap(_.listFiles).filter(_.isDirectory) // 00
```
Do we need to check the directory names too?

dongjoon-hyun replied:
Yes, it's a protected approach which excludes some random files in the intermediate level.

viirya replied:
Oh, will you add check directory name (executor-xxx, blockmgr-xxx, etc) later?

dongjoon-hyun replied:
Sure, we can check the pattern. I will do later~

viirya commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleExecutorComponents.scala:
```diff
@@ -0,0 +1,102 @@
+    // Find All files
+    val files = Utils.getConfiguredLocalDirs(conf)
+      .filter(_ != null)
+      .map(s => new File(new File(new File(s).getParent).getParent))
```
Why we need to get two-level parent? We store shuffle files outside local dirs?

dongjoon-hyun replied:
The current directory is the local directory for this executor. Here, we are trying to find dead executor's local directory. 
> We store shuffle files outside local dirs?

viirya commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleExecutorComponents.scala:
```diff
@@ -0,0 +1,102 @@
+        val decryptedSize = f.length()
+        bm.TempFileBasedBlockStoreUpdater(id, level, classTag, f, decryptedSize).save()
+      } catch {
+        case _: UnrecognizedBlockId =>
```
Is it good to ignore the exception? Looks like we will lose the block?

dongjoon-hyun replied:
No, `UnrecognizedBlockId` means the file name is not in Spark-generated pattern.
```scala
@DeveloperApi
class UnrecognizedBlockId(name: String)
    extends SparkException(s"Failed to parse $name into a block ID")
```

dongjoon-hyun commented:
Please let me know if there are any other comments, @viirya and @mridulm ?

viirya commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleDataIOSuite.scala:
```diff
@@ -0,0 +1,219 @@
+        .groupByKey()
+      rdd.collect()
+      assert(master.shuffleStatuses.keys.toSet == Set(0))
+      val loc1 = master.shuffleStatuses(0).mapStatuses(0).location
```
loc1 is not used?

dongjoon-hyun replied:
Yes, `loc1` is not used in this test case. I'll remove it.

viirya commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleDataIOSuite.scala:
```diff
@@ -0,0 +1,219 @@
+        assert(master.shuffleStatuses(1).mapStatuses.forall(_ == null))
+      }
+
+      rdd.collect()
+      assert(master.shuffleStatuses.keys.toSet == Set(0, 1))
+      assert(master.shuffleStatuses(0).mapStatuses.map(_.mapId).toSet == Set(17, 18, 19))
+      assert(master.shuffleStatuses(1).mapStatuses.map(_.mapId).toSet == Set(9, 10, 11, 12))
```
Why rdd is recomputed, but rdd2 is to recover from old shuffle status?

dongjoon-hyun replied:
This is a continuation of the previous test, `test("recompute is not blocked by the recovery")`.
We have two independent RDD `rdd` and `rdd2` and add is invoked first before recovery. `rdd2` is not used yet, but it's recovered.

viirya reviewed (approved):
looks good from my side.

dongjoon-hyun commented:
Thank you so much, @viirya ! Merged to master.

mridulm commented:
Late LGTM, apologies for the delay @dongjoon-hyun 
Thanks for reviewing @viirya !
